### DISCUSSION TITLE
Macos opening ghostty with open -n command opens it in a new process

### DISCUSSION THREAD

**Discussion by nadeemkhedr** [ghostty-org/ghostty#3563] [Q&A] [answered]
### Summary

I am using Karabiner to set some global shortcuts to launch new instances (windows) of applications

I have a shortcut that executes this command 
```sh
open -n "/Applications/Ghostty.app"
```

**What I expect:**
I expect this command to create a new window in ghostty

**What happens:**
This creates a new process (two ghostty apps running) and opens a new window 

### Why raise it as a question, not a bug?

docs for `-n` for the `open` command is:
> always open a new instance of application

But this behavior is different than other applications, I have been using this cmd with other applications like browsers and other terminals, and what happens is that it always open a new window not create a new process

**Comment by nadeemkhedr** [comment]
A workaround until/if this is fixed:

You can run this AppleScript:
```
tell application "Ghostty"
	if it is running then
		activate
		tell application "System Events" to keystroke "n" using {command down}
	else
		activate
	end if
end tell
```

This will open Ghostty if its closed, and will create a new window if it's open

**Reply by steveulfelder** [reply]
this works as expected to open the first terminal, and to open new windows in the workspace ghostty is running in.

however, when opening a new terminal in a different workspace, this script switches to the workspace ghostty is already running in (`if it is running then activate`) and opens the new terminal in the original ghostty's workspace, which is not the expected/desired behavior. an alternative:

```
tell application "Ghostty"
    if it is running
        tell application "System Events" to tell process "Ghostty"
            click menu item "New Window" of menu "File" of menu bar 1
        end tell
    else
        activate
    end if
end tell
```

**Reply by leviouwendijk** [reply]
Was looking for this exactly. Avoids workspace interference.

**Comment by 00-kat** [comment]
> browsers

I'm not 100% sure about macOS, but in my experience, Firefox doesn't support opening multiple instances, which may be relevant (of course, I don't know what browser you use; I'm just mentioning this in case you use Firefox).

**Comment by timvisher** [comment]
> But this behavior is different than other applications, I have been using this cmd with other applications like browsers and other terminals, and what happens is that it always open a new window not create a new process

This is the expected behavior of `open … -n`, IIUC. Chrome specifically is an odd-ball in this arena since it's tabs are all separate processes which is why it works the way you're describing.

I'm not sure which terminal you're talking about but Terminal.app, iTerm 2, Kitty, and Alacritty all open multiple application instances on my box.

When using this sort of logic in the future you probably want to switch to the standard

```
open -a <Application Name> -n
```

rather than what you're doing as well. Not only will this find Applications not installed in `/Applications` but it's also ever so slightly shorter.

**Reply by canTuncPekkan** [reply]
nice catch, I missed it by thinking the flag `-n` was a `-a`

**Reply by nadeemkhedr** [reply]
`open -a "Ghostty.app" -n` still opens ghostty as a new instance for me

**Reply by timvisher** [reply]
> `open -a "Ghostty.app" -n` still opens ghostty as a new instance for me

Sorry to be clear, that's what's supposed to happen. As you point out in OP

> docs for `-n` for the `open` command is:
> 
> > always open a new instance of application

The only application I'm currently aware of that doesn't behave this way is Google Chrome and that's only because Google Chrome runs every tab as a separate process (this was one of its big selling points in the beginning because a crash in one tab didn't bring down the whole browser).

As I pointed out:

> I'm not sure which terminal you're talking about but Terminal.app, iTerm 2, Kitty, and Alacritty all open multiple application instances on my box.

So I'm not sure what other terminals you're testing in but at least for me the behavior matches other terminals and what macOS claims `open … -n` should do.

**Reply by nadeemkhedr** [reply]
@timvisher 

Most of the terminals I used had options for this 

`WezTerm` -> By default works as a single instance only
`Kitty` -> you can add `--single-instance` options and would work like that 

Even Ghostty has the option to run a [single instance](https://ghostty.org/docs/config/reference#gtk-single-instance) only on GTK

I don't know if the correct solution is to add a single instance flag, for me this is not a blocker because i was able to find a workaround with the applescript, but if I couldn't it would have been one

**Reply by timvisher** [reply]
Ah so `Kitty` requires a special flag to inform it not to launch a separate process if one's running already I guess.

WezTerm may be taking the same approach that Google Chrome is using. I haven't tried that terminal at all.

> Even Ghostty has the option to run a [single instance](https://ghostty.org/docs/config/reference#gtk-single-instance) only on GTK

I think what's important here is that that's the default behavior on most linux distros I've seen which is why the gtk option makes sense there. If you run an application again in linux distros you tend to get a separate process. In those contexts Chrome is again the odd ball in that each new process becomes a separate tab.

Regardless it's not the default behavior on macOS and IMO applications that violate that expectation should not be emulated in an application that seeks to feel like a native app.

**Comment by timvisher** [comment]
Also you may want to follow https://github.com/ghostty-org/ghostty/discussions/2353 since I think a direct way to manage Ghostty GUI components is what you're looking for rather than for `open -a Ghostty -n` to do what you're asking it do.

**Comment by gknops** [comment]
For Terminal.app I have this handy function in my fish config:

```
function n
	
	echo "
tell application \"Terminal\"
	activate
	do script with command \"$argv\"
end tell
" | osascript

end

```

That allows me to write `n someCommand arg1 arg2` and have that happen in a fresh window. Would love to see this for Ghostty.

**Comment by paolomainardi** [comment]
Following this discussion because I am suffering the same issue with iterm2 and skhd I used to have this config:

```shell
❯ cat .config/skhd/skhdrc

cmd - return : osascript -e 'tell application "iTerm2" to create window with default profile command ""' || open -na iTerm
cmd + shift - return : open -n -a "Google Chrome"
```

And that osascript allows to open new windows in the same instance, instead of creating multiple iterm2 instances.

**Comment by mitchellh** [comment, marked as answer]
A new instance is the intended behavior of `-n` as noted here: https://github.com/ghostty-org/ghostty/discussions/3563#discussioncomment-11699286 Other applications are behaving odd, not us. :) There's probably some other way to script this but otherwise will be #2353.

**Reply by paolomainardi** [reply]
Thank you for your reply, @mitchellh. I believe other apps allow this by prefixing the `osascript`, as shown in the example I reported here: https://github.com/ghostty-org/ghostty/discussions/3563#discussioncomment-11766359.

**Comment by enetx** [comment]
`alt - return : id=$(yabai -m query --windows | jq -re ".[] | select(.app == \"Ghostty\") | .id" | head -n 1) && [ -z "$id" ] && open -a Ghostty || (yabai -m window --focus $id && skhd -k "cmd - n")
`

**Reply by rodoxplo** [reply]
It works like a charm; thank you so much.

**Reply by vaygr** [reply]
This won't work on a multi-screen, right? It will open new window next to the existing Ghostty, not where the original focus was.

**Reply by matiasandrade** [reply]
anyone have a solution that works on aerospace? I've been having the minor annoyance of opening new instances, or when I try to create a new window with a script, it opens it on the workspace that already has ghostty windows instead of the current workspace.

ctrl-k = """exec-and-forget osascript -e 'tell application "Ghostty" to create window with default profile command ""' || open -na Ghostty"""

**Reply by difro** [reply]
I just use a custom bash script
```
#!/usr/bin/env bash

cur_workspace=$(aerospace list-workspaces --focused)

osascript -e '
tell application "Ghostty"
        if it is running then
                activate
                tell application "System Events" to keystroke "n" using {command down}
        else
                activate
        end if
end tell'

aerospace move-node-to-workspace --focus-follows-window "$cur_workspace"
```
and do this in aerospace.toml
```
alt-t = 'exec-and-forget ~/.config/aerospace/new_ghostty.sh'
```

**Reply by John-Lin** [reply]
> anyone have a solution that works on aerospace? I've been having the minor annoyance of opening new instances, or when I try to create a new window with a script, it opens it on the workspace that already has ghostty windows instead of the current workspace.
> 
> ctrl-k = """exec-and-forget osascript -e 'tell application "Ghostty" to create window with default profile command ""' || open -na Ghostty"""

This solution is simple, but it will result in multiple instances being launched.
```
cmd-enter = 'exec-and-forget open -n -a Ghostty'
```

**Reply by EzraEllette** [reply]
Thanks

**Reply by John-Lin** [reply]
Ghostty 1.3 ships native AppleScript. Just inline `tell application "Ghostty"` to new window directly in aerospace.toml.
```toml
    # open new Ghostty window in current workspace (uses Ghostty 1.3 native AppleScript)
    cmd-enter = '''exec-and-forget CUR_WS=$(aerospace list-workspaces --focused) &&
        osascript -e 'tell application "Ghostty"
            new window
            activate
        end tell' &&
        aerospace move-node-to-workspace --focus-follows-window "$CUR_WS"'''
```
